src/components/SQForm/SQForm.tsx

```diff
@@ -1,19 +1,50 @@
 import React from 'react';
-import PropTypes from 'prop-types';
 import Grid from '@material-ui/core/Grid';
-import {Formik, Form} from 'formik';
+import {GridProps} from '@material-ui/core';
+import {Formik, Form, FormikHelpers} from 'formik';
 import {useDebouncedCallback} from 'use-debounce';
 import * as Yup from 'yup';
 import {useInitialRequiredErrors} from '../../hooks/useInitialRequiredErrors';
 
-function SQForm({
+interface SQFormProps<Values> {
+  /** Form Input(s) */
+  children: React.ReactNode;
+  /** Bool to pass through to Formik. https://formik.org/docs/api/formik#enablereinitialize-boolean */
+  enableReinitialize?: boolean;
+  /** Form Entity Object */
+  initialValues: Values;
+  /** Any prop from https://material-ui.com/api/grid */
+  muiGridProps?: GridProps;
+  /**
+   * Form Submission Handler | @typedef onSubmit: (values: Values, formikBag: FormikBag) => void | Promise<any>
+   * IMPORTANT: If onSubmit is async, then Formik will automatically set isSubmitting to false on your behalf once it has resolved.
+   * This means you do NOT need to call formikBag.setSubmitting(false) manually.
+   * However, if your onSubmit function is synchronous, then you need to call setSubmitting(false) on your own.
+   *
+   * https://jaredpalmer.com/formik/docs/api/withFormik#handlesubmit-values-values-formikbag-formikbag--void--promiseany
+   * */
+  onSubmit: (
+    values: Values,
+    formikHelpers: FormikHelpers<Values>
+  ) => void | Promise<unknown>;
+  /**
+   * Yup validation schema shape
+   * https://jaredpalmer.com/formik/docs/guides/validation#validationschema
+   * */
+  validationSchema?: Record<
+    keyof Values,
+    Yup.AnySchema<Values[keyof Values] | null | undefined>
+  >;
+}
+
+function SQForm<Values>({
```

Comment:
Hm. I've just seen this. Why does SQForm need an argument like this? This is definitely at the limit of my TS knowledge.

As a consumer what would I pass here?

Reply:
I think it's for the props `initialValues` and `validationSchema` to be able to use that to define their types.

Reply:
cool, I just watched cody's video above.

So when i consume then I don't need to `SQForm<someTypehere>` , TS will infer that type based on what I pass into `initialValues`?

Or do i need a seperate interface, `initialValuesTypes` outlining the types of the my initialValues?

Reply:
It should get the type from whatever you put in `initialValues`. I'm going to take a second look at this because we may not need to have that generic on `SQForm`, which could prevent any future confusion about needing to provide a generic

Reply:
Oh I didn't know it inferred the values that's awesome.

Reply:
It was new for me, this thing is awesome, I didn't we could pull off something like this using inferring type

Reply:
Alright, I took another look and I do think this is the way to go. With the `Values` generic on SQForm, we infer the type from `initialValues` and pass it to the `SQFormProps` interface so we strongly type the `onSubmit` and `validationSchema` props. You could still provide the `<SQForm>` component with a value for that generic, but `initialValues` is a required prop and linked to that generic, so whatever you define in the generic will have to match the type of whatever is given to `initialValues`. However, there's no added value in doing that -- the consumer won't need to know or care about the generic

[This SO post](https://stackoverflow.com/questions/67869151/infer-generic-type-between-2-props-in-react-typescript) shows the same solution in a more simple context -- that may better illustrate what's going on here if anyone is confused